File: 22-1-2025_Add Number Linked Lists.py
class Solution:
    def reverse(self, head):
        prev = None
        current = head
        next = None

        while current is not None:
            next = current.next
            current.next = prev
            prev = current
            current = next

        return prev

    def addTwoLists(self, num1, num2):
        num1 = self.reverse(num1)
        num2 = self.reverse(num2)

        sumList = None
        carry = 0

        while num1 is not None or num2 is not None or carry > 0:
            newVal = carry

            if num1 is not None:
                newVal += num1.data
            if num2 is not None:
                newVal += num2.data

            carry = newVal // 10
            newVal = newVal % 10

            newNode = Node(newVal)
            newNode.next = sumList
            sumList = newNode

            if num1:
                num1 = num1.next
            if num2:
                num2 = num2.next

        while sumList is not None and sumList.data == 0:
            temp = sumList.next
            sumList.next = None
            sumList = temp

        if sumList is None:
            return Node(0)
        return sumList

File: 22-1-2025_Add Number Linked Lists.cpp
class Solution {
  public:
    Node* addTwoLists(Node* num1, Node* num2) {
        Node* l1 = reverseList(num1);
        Node* l2 = reverseList(num2);
        Node* curr = nullptr;
        int carry = 0;

        while (l1 != nullptr || l2 != nullptr || carry > 0) {
            int v1 = (l1 != nullptr) ? l1->data : 0;
            int v2 = (l2 != nullptr) ? l2->data : 0;
            int sum = v1 + v2 + carry;
            carry = sum / 10;
            Node* tmp = new Node(sum % 10);
            tmp->next = curr;
            curr = tmp;

            if (l1 != nullptr) l1 = l1->next;
            if (l2 != nullptr) l2 = l2->next;
        }

        while (curr != nullptr && curr->data == 0) {
            curr = curr->next;
        }

        return curr;
    }

  private:
    Node* reverseList(Node* head) {
        Node* tmpHead = nullptr;

        while (head != nullptr) {
            Node* nex = head->next;
            if (tmpHead == nullptr) {
                tmpHead = head;
                tmpHead->next = nullptr;
            } else {
                head->next = tmpHead;
                tmpHead = head;
            }
            head = nex;
        }

        return tmpHead;
    }
};
